pkg/oramExecutor/oramExecutor.go

```diff
@@ -19,64 +21,138 @@ const (
 	stashSize   = 100000 // Maximum number of blocks in stash
 )
 
+type Operation struct {
+	RequestID uint64
+	Key       string
+	Value     string
+	Index     int
+}
+
+type ResponseTracker struct {
+	Values    []string
+	Count     int32 // Use atomic operations
+	Completed chan struct{}
+}
+
 type MyOram struct {
 	executor.UnimplementedExecutorServer
-	// rdb *redis.Client
 	o *ORAM
+
+	opMutex sync.Mutex
+	opCond  *sync.Cond
+	opQueue []Operation
+
+	trackerMutex sync.Mutex
+	trackers     map[uint64]*ResponseTracker
+
+	requestID uint64 // atomic
+	batchSize int
 }
 
 type StringPair struct {
 	First  string
 	Second string
 }
 
-func (e MyOram) ExecuteBatch(ctx context.Context, req *executor.RequestBatchORAM) (*executor.RespondBatchORAM, error) {
-	fmt.Printf("Got a request with ID: %d \n", req.RequestId)
+func (e *MyOram) ExecuteBatch(ctx context.Context, req *executor.RequestBatchORAM) (*executor.RespondBatchORAM, error) {
+	if len(req.Keys) != len(req.Values) {
+		return nil, fmt.Errorf("keys and values length mismatch")
+	}
 
-	// set batchsize
-	batchSize := 60
+	// Generate unique request ID
+	id := atomic.AddUint64(&e.requestID, 1)
+	numOps := len(req.Keys)
 
-	// Batching(requests []request.Request, batchSize int)
+	// Setup response tracker
+	tracker := &ResponseTracker{
+		Values:    make([]string, numOps),
+		Count:     0,
+		Completed: make(chan struct{}),
+	}
 
-	var replyKeys []string
-	var replyVals []string
+	// Register tracker
+	e.trackerMutex.Lock()
+	e.trackers[id] = tracker
+	e.trackerMutex.Unlock()
+
+	// Queue operations
+	e.opMutex.Lock()
+	for i := 0; i < numOps; i++ {
+		e.opQueue = append(e.opQueue, Operation{
+			RequestID: id,
+			Key:       req.Keys[i],
+			Value:     req.Values[i],
+			Index:     i,
+		})
+	}
+	e.opCond.Signal() // Notify batch processor
+	e.opMutex.Unlock()
```

Comment:
You can use a channel as a channel (maybe even a buffered channel) instead of using a list. I think each executor's queue in bacher is a channel and not a list. You can keep checking the size of the channel to know when its upto batch size.